### DISCUSSION TITLE
Imaginary PDF Rasterisation

### DISCUSSION THREAD

**Discussion by armouredking** [nextcloud/all-in-one#1906] [Questions] [answered]
I noticed that PDF is not one of the preview types used by the current Imaginary docker published by AIO. PDF should be something Imaginary can do. Is this a compile limitation or a setup one? Any thoughts to allowing it to do so, particularly over ImageMagick?

**Comment by szaimen** [comment]
Hi, xou could try to add pdf here: https://github.com/nextcloud/server/blob/e81cb6eaf0274dcdf2402de16216e8c61826235e/lib/private/Preview/Imaginary.php#L60. If that works, we could add pdf officially in order to add support for it.

**Reply by szaimen** [reply]
The mimetype of pdf should be this one: https://github.com/nextcloud/server/blob/e81cb6eaf0274dcdf2402de16216e8c61826235e/resources/config/mimetypemapping.dist.json#L146

**Reply by armouredking** [reply]
I attempted to do this by changing the line from:
`return '/image\/(bmp|x-bitmap|png|jpeg|gif|heic|svg|webp)/';`
to this:
`return '/image\/(bmp|x-bitmap|png|jpeg|gif|heic|svg|webp)\/application\/pdf/';`
But this results in:
`Error: preg_match(): Unknown modifier 'a' at /srv/data/html/nextcloud/lib/private/Preview/Generator.php#148`

Does it need to be an array instead of a string? Because we're including both image and application mimetypes?

**Reply by armouredking** [reply]
I went through looking at the other files in `lib/private/Preview` but didn't see any examples of multiple mimetypes. Seems all the "generators" are built for a single use case, so not sure how to define the line. I'm assuming based on looking at the other files it's RegXP of some sort, but not sure what separator it needs, or, if above it needs an array on the function.

**Reply by armouredking** [reply]
On a whim (even though it's ugly) I tried
`return '/(application|image)\/(bmp|x-bitmap|png|jpeg|gif|heic|pdf|svg|webp)/';`
which gave me the following:
```
[no app in context] Error: GuzzleHttp\Exception\ClientException: Client error: `POST http://127.0.0.1:9000/pipeline?operations=%5B%7B%22operation%22%3A%22autorotate%22%7D%2C%7B%22operation%22%3A%22smartcrop%22%2C%22params%22%3A%7B%22width%22%3A256%2C%22height%22%3A256%2C%22stripmeta%22%3A%22true%22%2C%22type%22%3A%22jpeg%22%2C%22norotation%22%3A%22true%22%7D%7D%5D` resulted in a `400 Bad Request` response:
{"message":"Error while processing the image: Unsupported image output type","status":400}
 at <<closure>>

 0. /srv/data/html/nextcloud/3rdparty/guzzlehttp/guzzle/src/Middleware.php line 69
    GuzzleHttp\Exception\RequestException::create()
 1. /srv/data/html/nextcloud/3rdparty/guzzlehttp/promises/src/Promise.php line 204
    GuzzleHttp\Middleware::GuzzleHttp\{closure}("*** sensitive parameters replaced ***")
 2. /srv/data/html/nextcloud/3rdparty/guzzlehttp/promises/src/Promise.php line 153
    GuzzleHttp\Promise\Promise::callHandler()
 3. /srv/data/html/nextcloud/3rdparty/guzzlehttp/promises/src/TaskQueue.php line 48
    GuzzleHttp\Promise\Promise::GuzzleHttp\Promise\{closure}("*** sensitive parameters replaced ***")
 4. /srv/data/html/nextcloud/3rdparty/guzzlehttp/promises/src/Promise.php line 248
    GuzzleHttp\Promise\TaskQueue->run()
 5. /srv/data/html/nextcloud/3rdparty/guzzlehttp/promises/src/Promise.php line 224
    GuzzleHttp\Promise\Promise->invokeWaitFn()
 6. /srv/data/html/nextcloud/3rdparty/guzzlehttp/promises/src/Promise.php line 269
    GuzzleHttp\Promise\Promise->waitIfPending()
 7. /srv/data/html/nextcloud/3rdparty/guzzlehttp/promises/src/Promise.php line 226
    GuzzleHttp\Promise\Promise->invokeWaitList()
 8. /srv/data/html/nextcloud/3rdparty/guzzlehttp/promises/src/Promise.php line 62
    GuzzleHttp\Promise\Promise->waitIfPending()
 9. /srv/data/html/nextcloud/3rdparty/guzzlehttp/guzzle/src/Client.php line 187
    GuzzleHttp\Promise\Promise->wait()
10. /srv/data/html/nextcloud/lib/private/Http/Client/Client.php line 289
    GuzzleHttp\Client->request()
11. /srv/data/html/nextcloud/lib/private/Preview/Imaginary.php line 112
    OC\Http\Client\Client->post()
12. /srv/data/html/nextcloud/lib/private/Preview/GeneratorHelper.php line 63
    OC\Preview\Imaginary->getCroppedThumbnail()
13. /srv/data/html/nextcloud/lib/private/Preview/Generator.php line 272
    OC\Preview\GeneratorHelper->getThumbnail()
14. /srv/data/html/nextcloud/lib/private/Preview/Generator.php line 151
    OC\Preview\Generator->getSmallImagePreview()
15. /srv/data/html/nextcloud/lib/private/Preview/Generator.php line 114
    OC\Preview\Generator->generatePreviews()
16. /srv/data/html/nextcloud/lib/private/PreviewManager.php line 185
    OC\Preview\Generator->getPreview()
17. /srv/data/html/nextcloud/core/Controller/PreviewController.php line 144
    OC\PreviewManager->getPreview()
18. /srv/data/html/nextcloud/core/Controller/PreviewController.php line 113
    OC\Core\Controller\PreviewController->fetchPreview()
19. /srv/data/html/nextcloud/lib/private/AppFramework/Http/Dispatcher.php line 225
    OC\Core\Controller\PreviewController->getPreviewByFileId()
20. /srv/data/html/nextcloud/lib/private/AppFramework/Http/Dispatcher.php line 133
    OC\AppFramework\Http\Dispatcher->executeController()
21. /srv/data/html/nextcloud/lib/private/AppFramework/App.php line 172
    OC\AppFramework\Http\Dispatcher->dispatch()
22. /srv/data/html/nextcloud/lib/private/Route/Router.php line 298
    OC\AppFramework\App::main()
23. /srv/data/html/nextcloud/lib/base.php line 1047
    OC\Route\Router->match()
24. /srv/data/html/nextcloud/index.php line 36
    OC::handleRequest()

GET /core/preview?fileId=431&c=308e8a813aa9174a07b6d7895fe335ab&x=250&y=250&forceIcon=0&a=0
from ##ip## by ##username## at 2023-02-03T06:43:47+00:00
```

Probably a cleaner way to do that line of code, but looks like it'll need more setup to understand the output.

**Reply by szaimen** [reply]
Hm... After looking a swcond time at this, the current implementation may only allow image mimetypes.

**Reply by szaimen** [reply]
Can you check if this works?
`return '/(image\/(bmp|x-bitmap|png|jpeg|gif|heic|svg|tiff|webp)|application\/pdf))/';`

**Reply by armouredking** [reply]
I tried that, but I think it had too many closing parenthetical on the end.
```
[PHP] Error: Error: preg_match(): Compilation failed: unmatched closing parenthesis at offset 67 at /srv/data/html/nextcloud/lib/private/Preview/Generator.php#326 at <<closure>>

 0. <<closure>>
    OC\Log\ErrorHandler::onError()
 1. /srv/data/html/nextcloud/lib/private/Preview/Generator.php line 326
    preg_match()
 2. /srv/data/html/nextcloud/lib/private/Preview/Generator.php line 162
    OC\Preview\Generator->getMaxPreview()
 3. /srv/data/html/nextcloud/lib/private/Preview/Generator.php line 114
    OC\Preview\Generator->generatePreviews()
 4. /srv/data/html/nextcloud/lib/private/PreviewManager.php line 185
    OC\Preview\Generator->getPreview()
 5. /srv/data/html/nextcloud/core/Controller/PreviewController.php line 144
    OC\PreviewManager->getPreview()
 6. /srv/data/html/nextcloud/core/Controller/PreviewController.php line 113
    OC\Core\Controller\PreviewController->fetchPreview()
 7. /srv/data/html/nextcloud/lib/private/AppFramework/Http/Dispatcher.php line 225
    OC\Core\Controller\PreviewController->getPreviewByFileId()
 8. /srv/data/html/nextcloud/lib/private/AppFramework/Http/Dispatcher.php line 133
    OC\AppFramework\Http\Dispatcher->executeController()
 9. /srv/data/html/nextcloud/lib/private/AppFramework/App.php line 172
    OC\AppFramework\Http\Dispatcher->dispatch()
10. /srv/data/html/nextcloud/lib/private/Route/Router.php line 298
    OC\AppFramework\App::main()
11. /srv/data/html/nextcloud/lib/base.php line 1047
    OC\Route\Router->match()
12. /srv/data/html/nextcloud/index.php line 36
    OC::handleRequest()

GET /core/preview?fileId=433&c=50152b8d8436f0ae26cd0c340d3a5552&x=250&y=250&forceIcon=0&a=0
from ##ip## by ##username## at 2023-02-03T15:10:51+00:00
```
Tried
`return '/(image\/(bmp|x-bitmap|png|jpeg|gif|heic|svg|webp)|application\/pdf)/'`
as well, still getting the same 400 error for unsupported image type.

**Reply by szaimen** [reply]
hm, no idea then. this log, is it logged in imaginary? Then imaginary might actually not support pdf preview generation...

**Reply by armouredking** [reply]
Imaginary does, so long as the library (libvips) is built with that support (hence original ask about build for the AIO docker).
See: https://github.com/h2non/imaginary

>imaginary can read images from HTTP POST payloads, server local path or remote HTTP servers, supporting JPEG, PNG, WEBP, HEIF, and optionally TIFF, PDF, GIF and SVG formats if libvips@8.3+ is compiled with proper library bindings.

This log is from the Nextcloud one, but Imaginary provides a similar notice. It doesn't have the Nextcloud specific file trace, just the trace for error `ErrOutputFormat` leading to `http.StatusBadRequest` or Error 400 as was similarly noted for `ErrResolutionTooBig` in [#409](https://github.com/h2non/imaginary/issues/409).

Example:
`172.21.0.1 - - [03/Feb/2023 15:07:21] "POST /pipeline?operations=%5B%7B%22operation%22%3A%22autorotate%22%7D%2C%7B%22operation%22%3A%22smartcrop%22%2C%22params%22%3A%7B%22width%22%3A256%2C%22height%22%3A256%2C%22stripmeta%22%3A%22true%22%2C%22type%22%3A%22jpeg%22%2C%22norotation%22%3A%22true%22%7D%7D%5D HTTP/1.1" 400 90 0.1230`

A "working" / "good" entry from Imaginary should be a code 200, like so:
`172.21.0.1 - - [03/Feb/2023 05:21:50] "POST /pipeline?operations=%5B%7B%22operation%22%3A%22autorotate%22%7D%2C%7B%22operation%22%3A%22fit%22%2C%22params%22%3A%7B%22width%22%3A4096%2C%22height%22%3A4096%2C%22stripmeta%22%3A%22true%22%2C%22type%22%3A%22jpeg%22%2C%22norotation%22%3A%22true%22%7D%7D%5D HTTP/1.1" 200 107011 0.0985`

Edit: changed quote to quote not code

**Reply by szaimen** [reply]
> 172.21.0.1 - - [03/Feb/2023 15:07:21] "POST /pipeline?operations=%5B%7B%22operation%22%3A%22autorotate%22%7D%2C%7B%22operation%22%3A%22smartcrop%22%2C%22params%22%3A%7B%22width%22%3A256%2C%22height%22%3A256%2C%22stripmeta%22%3A%22true%22%2C%22type%22%3A%22jpeg%22%2C%22norotation%22%3A%22true%22%7D%7D%5D HTTP/1.1" 400 90 0.1230

Hm... then it seems to be reached through to imaginary correctly but does not seem to work. So I guess your pdf example simply does not work or somehow imaginary is actually not able to generate previews of pdf.

**Reply by armouredking** [reply]
Could potentially be related to [#115](https://github.com/h2non/imaginary/issues/115) as well, but I don't know...

**Reply by szaimen** [reply]
yes, looks like it. So we would need to force the type with the type parameter for svg and pdf? But anyway a server issue (enhancement).

**Comment by armouredking** [comment]
Hm. I tried an SVG from `https://dev.w3.org/SVG/tools/svgweb/samples/svg-files/cc.svg` and that does generate a preview as expected.

I think I'll try building from source and see if this is an issue with libvips or NC specifically.

**Reply by armouredking** [reply]
Alright @szaimen I figured this out.

Imaginary is returning an image type PDF, when we need it to return an actual image. I've done this now both with the current AIO image and a build from source, just to make double certain I really pegged it down (coding is not my strong suit)

Steps done:
1) Perform modification to Imaginary.php, line 60 as discussed previously. This will get you:
`return '/(image\/(bmp|x-bitmap|png|jpeg|gif|heic|svg|webp)|application\/pdf)/';`
2) Strip out all current operations so we only compare apples to apples. Do this by adding `//` to the lead of lines 94 to 106
3) Add a new operation like so:
```
[
  'operation' => 'convert',
  'params' => [ 'type' => 'jpeg', ]
]
```
4) Visit a folder with no previews and watch. Updates are very rapid and all PDFs should convert. I did note one blip, I have some PDF artbooks that are 300+ MB in size and those do not get passed, but I'm pretty sure that's another config option. It works just fine on the skeleton Nextcloud auto files.
5) Confirm in Imaginary docker log by:
`docker logs $imaginaryaiocontainername`
```
172.21.0.1 - - [03/Feb/2023 20:56:50] "POST /pipeline?operations=%5B%7B%22operation%22%3A%22convert%22%2C%22params%22%3A%7B%22type%22%3A%22jpeg%22%7D%7D%5D HTTP/1.1" 200 55231 0.1586
172.21.0.1 - - [03/Feb/2023 20:56:50] "POST /pipeline?operations=%5B%7B%22operation%22%3A%22convert%22%2C%22params%22%3A%7B%22type%22%3A%22jpeg%22%7D%7D%5D HTTP/1.1" 200 55231 0.1427
172.21.0.1 - - [03/Feb/2023 20:56:50] "POST /pipeline?operations=%5B%7B%22operation%22%3A%22convert%22%2C%22params%22%3A%7B%22type%22%3A%22jpeg%22%7D%7D%5D HTTP/1.1" 200 45255 0.1965
172.21.0.1 - - [03/Feb/2023 20:56:50] "POST /pipeline?operations=%5B%7B%22operation%22%3A%22convert%22%2C%22params%22%3A%7B%22type%22%3A%22jpeg%22%7D%7D%5D HTTP/1.1" 200 45255 0.1958
172.21.0.1 - - [03/Feb/2023 20:56:50] "POST /pipeline?operations=%5B%7B%22operation%22%3A%22convert%22%2C%22params%22%3A%7B%22type%22%3A%22jpeg%22%7D%7D%5D HTTP/1.1" 200 45255 0.1957
172.21.0.1 - - [03/Feb/2023 20:56:50] "POST /pipeline?operations=%5B%7B%22operation%22%3A%22convert%22%2C%22params%22%3A%7B%22type%22%3A%22jpeg%22%7D%7D%5D HTTP/1.1" 200 81357 0.5798
```

![image](https://user-images.githubusercontent.com/16038428/216710029-9c5789c4-b02f-4c56-9e76-685cc4f4a9bf.png)

Obviously my code is trash and isn't for production (given it would just convert everything to jpegs), I think we probably want some of those parameters getting passed to Imaginary for PDF previews but this is just for proof-of-theory, and that's what we need to do to get PDFs working. Would this be an option on nextcloud/server for enhancement since all the required edits are in Imaginary.php?

**Reply by armouredking** [reply]
Picture of the full code
![image](https://user-images.githubusercontent.com/16038428/216710720-0da5b78b-61e6-4ae9-bd73-3f2482948022.png)

**Comment by armouredking** [comment, marked as answer]
Going to close this with opening `nextcloud/server` issue [#36543](https://github.com/nextcloud/server/issues/36543).
